### DISCUSSION TITLE
Tempered posterior in Blackjax

### DISCUSSION THREAD

**Discussion by Davidmolin4** [blackjax-devs/blackjax#369] [Q&A] [answered]
Hi everyone,

I would like to know if it is possible to implement the "log tempered posterior" which is defined as follows:

$logp(\theta | x, y)^{\frac{1}{T}} = \frac{1}{T} \left( logp(y|\theta, x) + logp(\theta) \right) + C(T),$

with $T>0$ and $C(T)$ a normalization constant.

I don't know if it is necessary to find the constant $C(T)$ otherwise it is straightforward.

**Comment by rlouf** [comment]
Hi,

Would you mind giving me more background: a paper, an example, etc. so I can better answer your question ?

**Reply by Davidmolin4** [reply]
In section two of this [paper](https://arxiv.org/pdf/2102.06571.pdf) discusses more about :)

**Reply by Davidmolin4** [reply]
In other words, instead of sampling from 

$p(\theta| \mathcal{D})$

I would like to sample from 

$\frac{p(\theta| \mathcal{D})^{\frac{1}{T}}}{\int p(\theta| \mathcal{D})^{\frac{1}{T}} d\theta}$

with $T>0$

**Reply by rlouf** [reply]
Thank you. I will read a little and come back to you in the next couple of days.

**Reply by Davidmolin4** [reply]
Thanks!

**Reply by rlouf** [reply]
Excuse me for the delay but you should be able to sample from a tempered posterior using SgLD by rescaling the gradient by the temperature. Look at the way rescaling is done in `csgld.py` in #396. You would only have to keep the terms related to the temperature (it should be easy to derive the expression based on the equations of the Langevin dynamics).

**Reply by Davidmolin4** [reply]
No problem! Thank you for your reply.

I could not understand how you implemented it :/ 
Is it possible to simply rescale the logprior and loglikelihood by temperature? I did so, but the model accuracy is equal to 100/(number of classes) so something tells me it's wrong, for temperatures lower than 0.1.

**Reply by rlouf** [reply]
You can't simply rescale because changing the functional form of your potential changes the equations of motion! If you simply rescale, the Langevin dynamics that is used to draw samples does not target your original distribution anymore but something else, hence the results you see.

It's a good exercise for me too, so I'll take a look at the theory again and come back to you with an explanation, and hopefully a solution.

**Reply by Davidmolin4** [reply]
Ok! Thank you for your support

**Reply by rlouf** [reply]
You can follow the progress on https://github.com/blackjax-devs/blackjax/pull/396 where we're adding support for temperature != 1 in SgLD and SgHMC.

**Reply by Davidmolin4** [reply]
Thanks for the information :)

**Reply by rlouf** [reply]
You can now use a temperature != 1 with both SGLD and SGHMC with `blackjax-nightly`.

**Comment by rlouf** [comment, marked as answer]
You can now use a temperature != 1 with both SGLD and SGHMC with blackjax-nightly.
